npm/private/npm_translate_lock.bzl

```diff
@@ -506,21 +497,72 @@ Check the public_hoist_packages attribute for duplicates.
                 fail(msg)
 
 def _impl(rctx):
-    lockfile = _process_lockfile(rctx)
+    lockfile = None
+    root_package = None
+    link_workspace = None
+    lockfile_description = None
+
+    if rctx.attr.pnpm_lock != None:
+        if rctx.attr.npm_package_json != None or rctx.attr.npm_package_lock != None:
+            fail("npm_translate_lock requires either pnpm_lock or npm_package_json/npm_package_lock attributes, not both.")
+
+        # don't allow a pnpm lock file that isn't in the root directory of a bazel package
+        if paths.dirname(rctx.attr.pnpm_lock.name):
+            fail("pnpm-lock.yaml file must be at the root of a bazel package")
 
-    # root package is the directory of the pnpm_lock file
-    root_package = rctx.attr.pnpm_lock.package
+        lockfile = _process_lockfile(rctx, rctx.attr.pnpm_lock)
 
-    generated_by_lines = _generated_by_lines(rctx.attr.pnpm_lock.workspace_name, rctx.attr.pnpm_lock)
+        # root package is the directory of the pnpm_lock file
+        root_package = rctx.attr.pnpm_lock.package
+        link_workspace = rctx.attr.pnpm_lock.workspace_name
+
+        lockfile_description = "@{}{}".format(
+            str(rctx.attr.pnpm_lock.workspace_name),
+            str(rctx.attr.pnpm_lock),
+        )
+    else:
+        if rctx.attr.npm_package_json == None or rctx.attr.npm_package_lock == None:
+            fail("npm_translate_lock requires either pnpm_lock or npm_package_json/npm_package_lock attributes.")
+
+        rctx.file(
+            "package.json",
+            content = rctx.read(rctx.attr.npm_package_json),
```

Comment:
Looks like this is what is read from it: https://github.com/pnpm/pnpm/blob/2a34b21ced6e4d033659405cec63b182649e8238/packages/plugin-commands-installation/src/getOptionsFromRootManifest.ts

I don't know exactly what all of these options do, but it seems reasonable to omit support for them and supply an empty package.json. Those that need them can fall back to manual `pnpm import` + `pnpm_lock` attribute.

Reply:
Nope, doesn't work. It reads other stuff too, and if package.json is empty it doesn't generate a pnpm-lock.yaml.

Sorry, I don't think there's a way around this.

Reply:
okay well there are a couple options:
- just document that when you use this variant, it's going to cause invalidations. It's not as bad as rules_nodejs - we won't re-run a whole `npm install` when it's invalidated, so maybe the perf loss isn't that noticeable
- get fancy and run `jq` on the package.json to derive one that only has the subset of bits pnpm reads - but I don't think this would work to avoid invalidations anyway since it's still one big repository rule. And it's brittle, pnpm could read more in their next version

Reply:
I documented that passing `package_json` causes invalidations.